extensions/data-transfer/portability-data-transfer-flickr/src/main/java/org/dataportabilityproject/datatransfer/flickr/photos/FlickrPhotosImporter.java

```diff
@@ -88,38 +89,50 @@ public ImportResult importItem(UUID jobId, AuthData authData, PhotosContainerRes
     }
 
-    for (PhotoAlbum album : data.getAlbums()) {
-      tempPhotosData.addAlbum(CACHE_ALBUM_METADATA_PREFIX + album.getId(), album);
+    if (data.getAlbums() != null) {
+      importAlbums(data.getAlbums(), tempPhotosData);
+      jobStore.update(jobId, tempPhotosData);
     }
-    jobStore.update(jobId, tempPhotosData);
 
-    for (PhotoModel photo : data.getPhotos()) {
+    if (data.getPhotos() != null) {
       try {
-        String photoId = uploadPhoto(photo);
-        String oldAlbumId = photo.getAlbumId();
-        TempPhotosData tempData = jobStore.findData(TempPhotosData.class, jobId);
-        String newAlbumId = tempData.lookupNewAlbumId(oldAlbumId);
-        if (Strings.isNullOrEmpty(newAlbumId)) {
-          // This means that we havent created the new album yet, create the photoset
-          PhotoAlbum album = tempData.lookupAlbum(CACHE_ALBUM_METADATA_PREFIX + oldAlbumId);
-          Photoset photoset =
-              photosetsInterface.create(
-                  COPY_PREFIX + album.getName(), album.getDescription(), photoId);
-          tempData.addAlbumId(oldAlbumId, photoset.getId());
-        } else {
-          // We've already created a new album, add the photo to the new album
-          photosetsInterface.addPhoto(newAlbumId, photoId);
-        }
-        jobStore.update(jobId, tempData);
+        importPhotos(jobId, data.getPhotos());
       } catch (FlickrException | IOException e) {
-        // TODO: figure out retries
-        return new ImportResult(
-            ImportResult.ResultType.ERROR, "Error importing item: " + e.getMessage());
+        return new ImportResult(ResultType.ERROR, "Error importing photo: " + e.getMessage());
       }
     }
+
     return new ImportResult(ImportResult.ResultType.OK);
   }
 
+  private void importAlbums(Collection<PhotoAlbum> albums, TempPhotosData tempPhotosData) {
+    for (PhotoAlbum album : albums) {
+      tempPhotosData.addAlbum(CACHE_ALBUM_METADATA_PREFIX + album.getId(), album);
+    }
+  }
+
+  private void importPhotos(UUID id, Collection<PhotoModel> photos)
+      throws FlickrException, IOException {
+    for (PhotoModel photo : photos) {
+      String photoId = uploadPhoto(photo);
+      String oldAlbumId = photo.getAlbumId();
+      TempPhotosData tempData = jobStore.findData(TempPhotosData.class, id);
+      String newAlbumId = tempData.lookupNewAlbumId(oldAlbumId);
+      if (Strings.isNullOrEmpty(newAlbumId)) {
+        // This means that we havent created the new album yet, create the photoset
+        PhotoAlbum album = tempData.lookupAlbum(CACHE_ALBUM_METADATA_PREFIX + oldAlbumId);
+        Photoset photoset =
+            photosetsInterface.create(
+                COPY_PREFIX + album.getName(), album.getDescription(), photoId);
+        tempData.addAlbumId(oldAlbumId, photoset.getId());
+      } else {
+        // We've already created a new album, add the photo to the new album
+        photosetsInterface.addPhoto(newAlbumId, photoId);
```

Comment:
Hmmm now that I think of it, it might not be the case that each photo is associated with an album - but we didn't account for that case in the old code. 

I guess a better way to do this would be to only add to an album if the photo is associated with an album. 

I have to think about and play with this case a bit more, so I added a TODO instead. 
Since in this case, we're assuming that each photo is associated with an album, the else should stay, since we add the photo when we create the photoset.